pkg/controller/pod.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	kubeClientSet "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+//Struct for the reconciler
+type ReconcilePod struct {
+	client                client.Client
+	kubeClient            kubeClientSet.Interface
+	azVolumeClient        azClientSet.Interface
+	namespace             string
+	controllerSharedState *SharedState
+}
+
+// Implement reconcile.Reconciler so the controller can reconcile objects
+var _ reconcile.Reconciler = &ReconcilePod{}
+
+func (r *ReconcilePod) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var pod corev1.Pod
+
+	podKey := getQualifiedName(request.Namespace, request.Name)
+	if err := r.client.Get(ctx, request.NamespacedName, &pod); err != nil {
+		// if the pod has been deleted, remove entry from podToClaimsMap and claimToPodMap
+		if errors.IsNotFound(err) {
+			r.controllerSharedState.deletePod(podKey)
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("failed to get Pod (%s): %v", podKey, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// update the shared map
+	r.controllerSharedState.addPod(&pod)
+
+	// create replica batches
+	if err := r.createReplicas(ctx, &pod); err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcilePod) createReplicas(ctx context.Context, pod *corev1.Pod) error {
+	if pod == nil {
+		return nil
+	}
+	podKey := getQualifiedName(pod.Namespace, pod.Name)
+	volumes, err := r.controllerSharedState.getVolumesFromPod(podKey)
+	if err != nil {
+		return err
+	}
+
+	v, _ := r.controllerSharedState.podLocks.LoadOrStore(pod, &sync.Mutex{})
+	podLock := v.(*sync.Mutex)
```

Comment:
I wonder if we should simply return if there is already a pending replica creation operation? Since creating the replicas will take non-trivial amount of time, a reconciler goroutine will end up being blocked for each Pod status update as it is brought to the running state. This could exhaust the reconciler's concurrency limit with even just a few Pods with replicas being created, and would lead to a storm of api server calls when goroutines are finally able to execute possibly resulting in throtting.

Reply:
Replica CRI creation itself isn't a time-consuming operation; the underlying attachment process is. Since, replica controller only acquires the lock to create the CRIs, it shouldn't be a huge blocker I don't think.

We could return and requeue the request but I don't think golang's `Mutex` allows checking for `Mutex`'s state. If we wanted to do this, I think we would need to use semaphores.

Reply:
Oh, right - the attach/detach controller handles the actual attachment process.

Reply:
I agree and I think for now we can address this situation or minimize its possibility by acquiring the lock before calling `addPod` on `SharedState` as in order for abovementioned scenario to be possible, pod entry would need to be present in `claimToPodsMap` and `podToClaimsMap`